bundle/edu.gemini.ags/src/main/scala/edu/gemini/ags/api/AgsHash.scala

```diff
@@ -0,0 +1,148 @@
+    * corresponding to the given time.
+    */
+  def hash(ctx: ObsContext, when: Instant): Int =
+    hash(ctx, when.toEpochMilli)
+
+
+  /** Calculates the AGS hash for this context, calculating the base position
+    * corresponding to the given time expressed in milliseconds since 1970.
+    */
+  def hash(ctx: ObsContext, when: Long): Int = {
+
+    val buf = ListBuffer.empty[Int]
+
+    // AGS Strategy
+    val strategyKey = AgsRegistrar.currentStrategy(ctx).map(_.key)
+    strategyKey.foreach { s =>
+      M3.stringHash(s.id) +=: buf
+    }
+
+    // Conditions
+    Option(ctx.getConditions).foreach { c =>
+      c.cc.## +=: c.iq.## +=: c.sb.## +=: buf
+    }
+
+    // Base Position
+    Option(ctx.getTargets).foreach { t =>
+      val time = Some(new java.lang.Long(when)).asGeminiOpt
+      val base = t.getBase.getTarget
+
+      def toData(coord: GemOption[java.lang.Double]): Int =
+        coord.asScalaOpt.map(_.doubleValue.scaled(8)).##
+
+      val ra  = toData(base.getRaDegrees(time))
+      val dec = toData(base.getDecDegrees(time))
+
+      ra +=: dec +=: buf
+    }
+
+    // Offset Positions, which are returned in a Set.  Order is not important
+    // for the purpose of AGS calculations.
+    M3.unorderedHash(ctx.getSciencePositions.asScala.map { o =>
+      (o.p.arcsec.scaled(3), o.q.arcsec.scaled(3))
+    }) +=: buf
+
+    // Position Angle
+    Option(ctx.getPositionAngle).foreach { a =>
+      a.degrees.scaled(3).## +=: buf
+    }
+
+    // Position Angle Constraint
+    Option(ctx.getPosAngleConstraint).foreach { pac =>
+      pac.## +=: buf
+    }
+
+    // IssPort
+    Option(ctx.getIssPort).foreach { iss =>
+      iss.## +=: buf
+    }
+
+    import AgsStrategyKey._
+
+    // Vignetting calculation. Instrument-specific features that have an impact
+    // on the science area and probe arm position and hence, vignetting.
+    Option(ctx.getInstrument).foreach {
+      case i: Flamingos2 if strategyKey.contains(Flamingos2OiwfsKey)   =>
+        i.getFpu.## +=: i.getLyotWheel.getPlateScale.## +=: buf
+
+      case i: InstGmosNorth if strategyKey.contains(GmosNorthOiwfsKey) =>
+        i.getFPUnit.## +=: i.getFPUnitMode.## +=: buf
+
+      case i: InstGmosSouth if strategyKey.contains(GmosSouthOiwfsKey) =>
+        i.getFPUnit.## +=: i.getFPUnitMode.## +=: buf
+
+      case _                                                           =>
+    }
+
+    // Strategy-specific differences.  In some instruments PWFS vignetting
+    // clearance changes based on instrument-specific features such as which
+    // camera is in use.
+    strategyKey.foreach {
+      case Pwfs1NorthKey | Pwfs1SouthKey =>
+        Option(ctx.getInstrument).foreach {
+          _.pwfs1VignettingClearance.getMagnitude.## +=: buf
+        }
+
+      case Pwfs2NorthKey | Pwfs2SouthKey =>
+        Option(ctx.getInstrument).foreach {
+          _.pwfs2VignettingClearance.getMagnitude.## +=: buf
+        }
+
+      case GemsKey                       =>
```

Comment:
GeMS is a little confusing because the code for the manual search offers many more options.  When an AGS computation is done, it fixes most of the options.  The things we need to include are those that differ based on the `ObsContext` alone.  For example, whether it is using GSAOI or F2.

It looked to me like the NIR band was being fixed at `None` when `search` is called in the AGS `select` method.  Since it is `None` the instrument in use will determine how the filtering is done, for example in the [`GemsMagnitudeTable`](https://github.com/gemini-hlsw/ocs/blob/develop/bundle/edu.gemini.ags/src/main/scala/edu/gemini/ags/gems/GemsMagnitudeTable.scala#L92).

I could easily be wrong though.  If you see where the NIR band is otherwise extracted from something in the context, let me know and I'll have to add that.